tests/unit/utils/test_utils.py

```diff
@@ -104,3 +105,28 @@ def test_relpath():
     path_info = PathInfo(path)
 
     assert relpath(path) == relpath(path_info)
+
+
+@pytest.mark.parametrize(
+    "inp,out,is_dir,expected",
+    [
+        ["target", None, False, "target"],
+        ["target", "dir", True, os.path.join("dir", "target")],
+        ["target", "file_target", False, "file_target"],
+        [
+            "target",
+            os.path.join("dir", "subdir"),
+            True,
+            os.path.join("dir", "subdir", "target"),
+        ],
+        ["dir/", None, False, "dir"],
+        ["dir", None, False, "dir"],
+        ["dir", "other_dir", False, "other_dir"],
+        ["dir", "other_dir", True, os.path.join("other_dir", "dir")],
+    ],
+)
+def test_resolve_output(inp, out, is_dir, expected):
```

Comment:
Personally, it's too hard for me to parse this matrix. I'd like to have it flattened-out somehow, eg:
```python
with mock.patch("os.path.isdir", return_value=True):
	assert resolve_output("target", "dir") == os.path.join("target", "dir")
    # ... more tests

with mock.patch("os.path.isdir", return_value=False):
	assert resolve_output("target", None) == "target"
    # ... more tests
```

You don't have to change anything, I'm just stating my preference. I'm fine with this as well.

Reply:
I agree that it might be easier to read through "flat" version, but this matrix (in my opinion) makes it easier to compare particular cases. Also, If we have "flat" version, during test execution, in case of singular fail. we will only know that this test failed, while in case of parameterization we will know exactly which argument set has failed.

Reply:
I would also say that many `with` statements one under another kind of obscures what the test is about.